src/layout.rs

```diff
@@ -1,6 +1,107 @@
 //! Final and cached data structures that represent the high-level UI layout
 
 use crate::geometry::{Point, Size};
+use crate::sys::abs;
+
+/// Whether we are performing a full layout, or we merely need to size the node
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum RunMode {
+    /// A full layout for this node and all children should be computed
+    PeformLayout,
+    /// The layout algorithm should be executed such that an accurate container size for the node can be determined.
+    /// Layout steps that aren't necessary for determining the container size of the current node can be skipped.
+    ComputeSize,
+}
+
+/// Whether styles should be taken into account when computing size
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum SizingMode {
+    /// Only content contributions should be taken into account
+    ContentSize,
+    /// Inherent size styles should be taken into account in addition to content contributions
+    InherentSize,
+}
+
+/// The amount of space available to a node in a given axis
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum AvailableSpace {
+    /// The amount of space available is the specified number of pixels
+    Definite(f32),
+    /// The amount of space available is indefinite and the node should be laid out under a min-content constraint
+    MinContent,
+    /// The amount of space available is indefinite and the node should be laid out under a max-content constraint
+    MaxContent,
+}
+
+impl AvailableSpace {
+    const ZERO: AvailableSpace = AvailableSpace::Definite(0.0);
+
+    /// Returns true for definite values, else false
+    pub fn is_definite(self) -> bool {
+        matches!(self, AvailableSpace::Definite(_))
+    }
+
+    /// Convert to Option
+    /// Definite values become Some(value). Contraints become None.
+    pub fn as_option(self) -> Option<f32> {
```

Comment:
I wonder if it would be better to implement `From<AvailableSpace>` for `Option<f32>` instead.

But maybe calling `.into::<Option<f32>>` instead of `.as_option` is too cubersome.

Reply:
I personally feel like as_option() reads clearer, as into could be converting to anything. Although perhaps it should be `.into_option`.

Reply:
Renamed to `into_option`